WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -44,30 +44,39 @@ public class CoinJoinClient
 		public Kitchen Kitchen { get; }
 		public KeyManager Keymanager { get; }
 		private RoundStateUpdater RoundStatusUpdater { get; }
+		public CoinJoinClientState State { get; private set; }
 
 		public async Task<bool> StartCoinJoinAsync(IEnumerable<SmartCoin> coins, CancellationToken cancellationToken)
 		{
-			var currentRoundState = await RoundStatusUpdater.CreateRoundAwaiter(roundState => roundState.Phase == Phase.InputRegistration, cancellationToken).ConfigureAwait(false);
+			State = CoinJoinClientState.InProgress;
+			try
+			{
+				var currentRoundState = await RoundStatusUpdater.CreateRoundAwaiter(roundState => roundState.Phase == Phase.InputRegistration, cancellationToken).ConfigureAwait(false);
 
-			// This should be roughly log(#inputs), it could be set slightly
-			// higher if more inputs are observed but that involves trusting the
-			// coordinator with those values. Therefore, conservatively set this
-			// so that a maximum of 5 blame rounds are executed.
-			// FIXME should smaller rounds abort earlier?
-			var tryLimit = 6;
+				// This should be roughly log(#inputs), it could be set slightly
+				// higher if more inputs are observed but that involves trusting the
+				// coordinator with those values. Therefore, conservatively set this
+				// so that a maximum of 5 blame rounds are executed.
+				// FIXME should smaller rounds abort earlier?
+				var tryLimit = 6;
 
-			for (var tries = 0; tries < tryLimit; tries++)
-			{
-				if (await StartRoundAsync(coins, currentRoundState, cancellationToken).ConfigureAwait(false))
-				{
-					return true;
-				}
-				else
+				for (var tries = 0; tries < tryLimit; tries++)
 				{
-					var blameRoundState = await RoundStatusUpdater.CreateRoundAwaiter(roundState => roundState.BlameOf == currentRoundState.Id, cancellationToken).ConfigureAwait(false);
-					currentRoundState = blameRoundState;
+					if (await StartRoundAsync(coins, currentRoundState, cancellationToken).ConfigureAwait(false))
+					{
+						return true;
+					}
+					else
+					{
+						var blameRoundState = await RoundStatusUpdater.CreateRoundAwaiter(roundState => roundState.BlameOf == currentRoundState.Id, cancellationToken).ConfigureAwait(false);
+						currentRoundState = blameRoundState;
+					}
 				}
 			}
+			finally
+			{
+				State = CoinJoinClientState.Idle;
```

Comment:
We could avoid this by querying only the `TrackedWallets` which `CoinJoinTask` is not completed. Basically, it makes no sense to ask for the state of a task that has been already completed. So, this state `Idle` is something artificial.

Reply:
I would feel safer if CoinJoinClient would set up its own state properly.

Reply:
it is not safer, it is redundant. It is like asking what color is my white horse. Or, "hey finished task, what are you doing?"

Reply:
I did not understand until now. I refactored CoinJoinClient.